Pyrado/pyrado/algorithms/meta/pddr.py

```diff
@@ -0,0 +1,363 @@
+        else:
+            self.load_teachers()
+            if self.num_teachers < len(self.teachers):
+                print(
+                    f"You have loaded {len(self.teachers)} teachers. Only the first {self.num_teachers} will be used!"
+                )
+                self.prune_teachers()
+        assert self.num_teachers == len(self.teachers)
+
+        # Student
+        self._expl_strat = NormalActNoiseExplStrat(self._policy, std_init=std_init)
+        self._policy = self._policy.to(self.device)
+        self.optimizer = to.optim.Adam([{"params": self.policy.parameters()}], lr=lr)
+
+        # Environments
+        self.env_real = env
+        self.rollouts = dict()
+        self.samplers = [
+            ParallelRolloutSampler(
+                self.teacher_envs[t],
+                deepcopy(self._expl_strat),
+                num_workers=int(self.num_cpu / self.num_teachers),
+                min_steps=self.min_steps,
+            )
+            for t in range(self.num_teachers)
+        ]
+
+        self.teacher_weights = np.ones(self.num_teachers)
+
+        # Criterion
+        self.criterion = to.nn.KLDivLoss(log_target=True, reduction="batchmean")
+
+    @property
+    def expl_strat(self) -> NormalActNoiseExplStrat:
+        return self._expl_strat
+
+    def step(self, snapshot_mode: str, meta_info: dict = None):
+        """
+        Perform a single iteration of the algorithm. This includes collecting the data, updating the parameters, and
+        adding the metrics of interest to the logger. Does not update the `curr_iter` attribute.
+
+        :param snapshot_mode: determines when the snapshots are stored (e.g. on every iteration or on new highscore)
+        :param meta_info: is not `None` if this algorithm is run as a subroutine of a meta-algorithm,
+                          contains a dict of information about the current iteration of the meta-algorithm
+        """
+        # Sample observations
+        ros = []
+        rost = []
+        for sampler in self.samplers:
+            samples = sampler.sample()
+            rost.append(samples)
+            ros.extend(samples)
+
+        rets = [ro.undiscounted_return() for ro in ros]
+        avg_ret = np.mean(rets)
+        all_lengths = np.array([ro.length for ro in ros])
+
+        # Log current progress
+        self.logger.add_value("max return", np.max(rets), 4)
+        self.logger.add_value("median return", np.median(rets), 4)
+        self.logger.add_value("avg return", np.mean(rets), 4)
+        self.logger.add_value("min return", np.min(rets), 4)
+        self.logger.add_value("std return", np.std(rets), 4)
+        self.logger.add_value("std var", self.expl_strat.std.item(), 4)
+        self.logger.add_value(
+            "avg rollout len", np.mean(all_lengths), 4
+        )  # the last used rollout is trimmed later to fit min_steps
+        # self.logger.add_value("num total samples", np.sum(all_lengths))        # is not what is used in update()
```

Comment:
Why is this commented out? I think the number of total samples of PDDR is the number of used samples of all its teachers and the student. This maybe needs to be calculated in a separate function (or not) but should be in the logger. (I am aware that some older algos dont have this, since it is rather new)